core/controllers/story_editor.py

```diff
@@ -59,6 +59,48 @@ class EditableStoryDataHandler(base.BaseHandler):
     """A data handler for stories which support writing."""
 
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    URL_PATH_ARGS_SCHEMAS = {
+        'story_id': {
+            'schema': {
+                'type': 'basestring'
+            },
+            'validators': [{
+                'id': 'has_length',
+                'value': constants.STORY_ID_LENGTH
+            }]
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {},
+        'PUT': {
+            'version': {
+                'schema': {
+                    'type': 'int',
+                    'default_value': 1
+                }
+            },
+            'commit_message': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_COMMIT_MESSAGE_LENGTH
+                    }]
+                }
+            },
+            'change_dicts': {
+                'schema': {
+                    'type': 'list',
+                    'items': {
+                        'type': 'object_dict',
+                        'validation_method': (
+                            story_domain.StoryChange)
```

Comment:
Hmm, I think it is a bit confusing to do `validation_method: story_domain.StoryChange`. Can `validate` do `to_dict` and then `validate_dict`? Also, in #14623 we are trying to solve the same thing so you can discuss this together if you want.

Reply:
I've already tried implementing that and I outlined what happened in the above comment's explanation.

Reply:
Sorry, I missed some stuff.

> I tried implementing a validate function using the logic you outlined, but I'm confused on where it would be called for BaseChange.

I don't think you need to call it from anywhere, it will be called by the code that already is in schema utils.

Reply:
I understand where `BaseChange` is called from during schema validation, but in the actual implementation of the class `BaseChange`, its initialization calls `validate_dict` immediately. You are suggesting that I change `validate_dict` to using `validate` alongside a `from_dict` call. The problem is that when I validate `StoryChange` in this context, the `BaseChange` hasn't even been initialized yet.

Reply:
@vojtechjelinek 

For reference, this is the `__init__` function for `BaseChange`:

```
def __init__(self, change_dict):
        """Initializes a BaseChange object from a dict.

        Args:
            change_dict: dict. The dict containing cmd name and attributes.

        Raises:
            ValidationError. The given change_dict is not valid.
        """
        self.validate_dict(change_dict)

        cmd_name = change_dict['cmd']
        self.cmd = cmd_name

        all_allowed_commands = (
            self.ALLOWED_COMMANDS + self.COMMON_ALLOWED_COMMANDS)

        cmd_attribute_names = []
        for cmd in all_allowed_commands:
            if cmd['name'] == cmd_name:
                cmd_attribute_names = (
                    cmd['required_attribute_names'] + cmd[
                        'optional_attribute_names'])
                break

        for attribute_name in cmd_attribute_names:
            setattr(self, attribute_name, change_dict.get(attribute_name))
```

Reply:
Also, you mentioned that `validation_method: story_domain.StoryChange` is confusing, but isn't that nearly identical to the alternative you're proposing?
> Hey, instead of doing this, can you introduce `from_dict` into the `BaseChange` that will create the object and then in the schema you can use `'object_class'` instead?
> 
> Also, can you do the same for `validate_exploration_change` and `validate_collection_change`?

`object_class` according to [Oppia's schema definition](https://github.com/oppia/oppia/wiki/Schemas) would be: `object_class: story_domain.StoryChange`.

Reply:
> I understand where `BaseChange` is called from during schema validation, but in the actual implementation of the class `BaseChange`, its initialization calls `validate_dict` immediately. You are suggesting that I change `validate_dict` to using `validate` alongside a `to_dict` and `validate_dict` call. The problem is that when I call `StoryChange` and `BaseChange` in this context, the parent class hasn't been initialized yet. `validate` and the following `to_dict` cannot be called on a non-existing object. Thus my initial confusion on where a `validate` method would be located.

I know that `validate_dict` is called in `__init__` but that doesn't really matter. It will validate it and create the object, then the schema utils will call `validate` again which is a bit redundant but it should be fine in this case. I do not suggest do modify `validate_dict` just adding a new `validate` function that will then call the already existing functions. This approach seems to be working in https://github.com/oppia/oppia/pull/14623/files#diff-c67a0705721f2177a9b6d5427a6146448949506b817f65293590710c5f077c2cR245


> Also, you mentioned that `validation_method: story_domain.StoryChange` is confusing, but isn't that nearly identical to the alternative you're proposing?
> 
> > Hey, instead of doing this, can you introduce `from_dict` into the `BaseChange` that will create the object and then in the schema you can use `'object_class'` instead?
> > Also, can you do the same for `validate_exploration_change` and `validate_collection_change`?
> 
> `object_class` according to [Oppia's schema definition](https://github.com/oppia/oppia/wiki/Schemas) would be: `object_class: story_domain.StoryChange`.

I mean it is different because when we use `validation_method` we expect a method or function to be passed, which in a technical sense `story_domain.StoryChange` also is but mainly it is a class, when we have `object_class` we expect a class so  `story_domain.StoryChange`  makes sense. If we did something like `object_class: service.get_class_abc` (say the function does `return AbcClass`) I would also say it is a bit antipattern and it should be refactored.